pycytominer/cyto_utils/cell_locations.py

```diff
@@ -0,0 +1,403 @@
+
+            if not self.single_cell_input.endswith(".sqlite"):
+                raise ValueError("single_cell file must be a SQLite file")
+
+            # if the single_cell file is an S3 path, download it to a temporary file
+            if self.single_cell_input.startswith("s3://"):
+                temp_single_cell_input = self._download_s3(self.single_cell_input)
+
+                # connect to the single_cell file
+                conn = sqlite3.connect(temp_single_cell_input)
+
+            else:
+                # connect to the single_cell file
+                conn = sqlite3.connect(self.single_cell_input)
+        else:
+            conn = self.single_cell_input
+
+        # Verify that the Image and Nuclei tables are present in single_cell
+
+        c = conn.cursor()
+
+        c.execute("SELECT name FROM sqlite_master WHERE type='table';")
+
+        tables = c.fetchall()
+
+        tables = [x[0] for x in tables]
+
+        if not ("Image" in tables and "Nuclei" in tables):
+            raise ValueError(
+                "Image and Nuclei tables are not present in the single_cell file"
+            )
+
+        # Verify that the required columns are present in the single_cell file
+
+        c.execute("PRAGMA table_info(Nuclei);")
+
+        nuclei_columns = c.fetchall()
+
+        nuclei_columns = [x[1] for x in nuclei_columns]
+
+        if not (
+            self.image_column in nuclei_columns
+            and self.object_column in nuclei_columns
+            and self.cell_x_loc in nuclei_columns
+            and self.cell_y_loc in nuclei_columns
+        ):
+            raise ValueError(
+                f"Required columns are not present in the Nuclei table in the SQLite file"
+            )
+
+        c.execute("PRAGMA table_info(Image);")
+
+        image_columns = c.fetchall()
+
+        image_columns = [x[1] for x in image_columns]
+
+        if not (
+            self.image_column in image_columns
+            and all(elem in image_columns for elem in self.image_index)
+        ):
+            raise ValueError(
+                f"Required columns are not present in the Image table in the SQLite file"
+            )
+
+        # Load the required columns from the single_cell file
+
+        nuclei_query = f"SELECT {self.image_column},{self.object_column},{self.cell_x_loc},{self.cell_y_loc} FROM Nuclei;"
+
+        image_index_str = ", ".join(self.image_index)
+
+        image_query = f"SELECT  {self.image_column},{image_index_str} FROM Image;"
+
+        nuclei_df = pd.read_sql_query(nuclei_query, conn)
+
+        image_df = pd.read_sql_query(image_query, conn)
+
+        conn.close()
+
+        # if the single_cell file was downloaded from S3, delete the temporary file
+        if "temp_single_cell_input" in locals():
+            pathlib.Path(temp_single_cell_input).unlink()
+
+        # Merge the Image and Nuclei tables
+        merged_df = pd.merge(image_df, nuclei_df, on=self.image_column, how="inner")
```

Comment:
Minor comment. Why is the merge done with pandas and not doing it directly in the SQL?. I'd expect faster execution in the latter.

Reply:
Because that's what Copilot made me do -- will ask Copilot to fix :D